### PR TITLE
Add standard values for generators

### PR DESCRIPTION
**Pull request by mhutch**
Add standard values in the project file property panel for generators (custom tools), based on:

1) Values that files in the project with the same extension have
2) Values that other files in the project have
3) Values handled by extensions

image

### CODE REVIEW

File: main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+			}
+
+			public override bool CanConvertTo (ITypeDescriptorContext context, Type destinationType)
+			{
+				return destinationType == typeof (string);
+			}
+
+			public override object ConvertTo (ITypeDescriptorContext context, System.Globalization.CultureInfo culture, object value, Type destinationType)
+			{
+				return (string)value;
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs]: Shouldn't this take culture into account?

File: main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+
+				string str = (string)value;
+				if (string.IsNullOrEmpty (str) || !char.IsLetter (str [0]))
+					return false;
+
+				for (int i = 1; i < str.Length; i++) {
+					char c = str [i];
+					if (char.IsLetterOrDigit (c) || c == '_')
+						continue;
+					else
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs]: redundant else.

File: main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.CustomToolStringsConverter.cs [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+				var values = new List<string> ();
+
+				var file = (context?.Instance as ProjectFileDescriptor)?.file;
+
+				//add sections with values that other files in the project have
+				//first, a section with values that files with this file's extension have
+				//then,  a section with  values that other files have
+				if (file?.Project != null) {
+					var extension = file.FilePath.Extension;
+					var defer = new HashSet<string> ();
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.CustomToolStringsConverter.cs]: Shouldn't this be ordinal ignore case too?

File: main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+			{
+				if (!IsValid (context, value))
+					throw new FormatException ("Invalid build target name");
+
+				return (string)value;
+			}
+
+			public override bool IsValid (ITypeDescriptorContext context, object value)
+			{
+				if (!(value is string))
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs]:
```suggestion
				if (!(value is string str))
```

And remove the local declaration below?

File: main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs (hunk 2 above)
Comment (mhutch) [reply main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs]: Existing code, did not modify, only moved

File: main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs (hunk 1 above)
Comment (mhutch) [reply main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.Projects/ProjectFileDescriptor.BuildActionStringsConverter.cs]: No, there's no meaningful way to change the culture of a build action string.

**Comment by Therzok** [comment]
Can we add some unit tests?

**Comment by mhutch** [comment]
Tests provided as requested :)

File: main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.csproj [hunk 5]
```diff
@@ -65,6 +65,9 @@
       <Private>False</Private>
     </ProjectReference>
   </ItemGroup>
+  <ItemGroup>
+    <InternalsVisibleTo Include="MonoDevelop.Ide.Tests" />
```
Comment (Therzok) [inline main/src/addins/MonoDevelop.DesignerSupport/MonoDevelop.DesignerSupport.csproj]: We need to add a new test assembly.

MonoDevelop.DesignerSupport depends on MonoDevelop.Ide, so those tests shouldn't technically run in a Ide.Tests environment.

**Comment by Therzok** [comment]
LGTM! I can help setting up the new test assembly and CI steps, as it's currently a layering violation test DesignerSupport in Ide.Tests.

**Comment by Therzok** [comment]
The environment now contains support for running [MonoDevelop.DesignerSupport.Tests.dll](https://devdiv.visualstudio.com/DevDiv/_releaseDefinition?definitionId=695&_a=definition-tasks&environmentId=2739) in the Release Definition.

File: main/tests/Ide.Tests/MonoDevelop.Ide.Tests.csproj [hunk 6]
```diff
@@ -193,6 +194,10 @@
       <Name>IdeUnitTests</Name>
       <Private>False</Private>
     </ProjectReference>
+    <ProjectReference Include="..\..\src\addins\MonoDevelop.DesignerSupport\MonoDevelop.DesignerSupport.csproj">
```
Comment (Therzok) [inline main/tests/Ide.Tests/MonoDevelop.Ide.Tests.csproj]: Layering violation here.

**Comment by Therzok** [comment]
Pushed 2 commits which should fix the layering violation and introduce a new test assembly.

**Comment by Therzok** [comment]
LGTM!

**Comment by Therzok** [comment]
cc @mrward

**Pull request merged**